Port of multilabel_margin_loss from TH to ATen (CPU)

andreaskoepf opened the pull request:
This is a port of the CPU version of TH MultiLabelMarginCriterion to ATen.

Benchmark results ([source of script used](https://gist.github.com/andreaskoepf/ce96eedb09e9480ae2263d31822ef26e)):

Slightly slower forward (probably acceptable), slightly faster forward & backward combination.

###  WITH patch:
```
CPU forward 1000 took 0.0002544010058045387
CPU forward 10000 took 0.0022866200015414506
CPU forward 100000 took 0.02240650000749156
CPU forward 1000000 took 0.22985397902084514
CPU forward 10000000 took 2.227811124001164
CPU forward TOTAL time 4.282580643019173
CPU for- & backward 1000 took 0.0006969539972487837
CPU for- & backward 10000 took 0.004804529016837478
CPU for- & backward 100000 took 0.07736711099278182
CPU for- & backward 1000000 took 0.5985556179948617
CPU for- & backward 10000000 took 4.761040163983125
CPU for- & backward TOTAL time 7.318476865999401
```

### WITHOUT patch:
```
CPU forward 1000 took 0.00026982801500707865
CPU forward 10000 took 0.002569925010902807
CPU forward 100000 took 0.024335263995453715
CPU forward 1000000 took 0.2151200629887171
CPU forward 10000000 took 2.114590842014877
CPU forward TOTAL time 4.184845258976566
CPU for- & backward 1000 took 0.0007158009975682944
CPU for- & backward 10000 took 0.005468863993883133
CPU for- & backward 100000 took 0.05931608600076288
CPU for- & backward 1000000 took 0.5732014369859826
CPU for- & backward 10000000 took 5.2500802429858595
CPU for- & backward TOTAL time 7.7646528169861995
```

andreaskoepf committed:
Port of multilabel_margin_loss from TH to ATen (CPU)

andreaskoepf committed:
Fix special case for 1d input

andreaskoepf commented:
@ezyang There is a strange inconsistency in handling 1d inputs between `torch.nn.functional.multi_margin_loss` and `torch.functional.multilabel_margin_loss` when the reduction argument is `none`. A unit test requires the former to return a scalar output in case of a 1d input and the latter wants a 1d output in case of a 1d input.

Output looks like this:
```
>>> torch.nn.functional.multi_margin_loss(torch.rand(10), torch.tensor([0]), reduction='none')
tensor(0.5028)
>>> torch.nn.functional.multilabel_margin_loss(torch.rand(4), torch.tensor([0,0,0,0]), reduction='none')
tensor([3.3168])
```

I also ported the behavior with scalar inputs ... here the output for multi_margin_loss and multilabel_margin_loss will always be zero and I think nobody is using it .. but since TH handles 1-element tensors in the same way as scalar inputs I copied this behavior.

andreaskoepf committed:
Remove old TH MultiLabelMarginCreterion implementation

ezyang commented:
Haven't had a chance to look at the inconsistency, but it looks like this needs rebasing :(

andreaskoepf committed:
Fix is_target parameter binding after TH CPU version removal

The argument istarget had to be renamed in function forward declarations
in THCUNN.h in order to allow compilation with CUDA enabled. There
is now a discrepancy between the arg name in the forward declaration
and the actual function definition but since the THCUNN function will
be removed in the future this might be acceptable.

andreaskoepf commented:
@ezyang rebased and hopefully fixed build, waiting for tests...

ezyang commented on aten/src/ATen/native/LossMultiLabelMargin.cpp:
```diff
@@ -0,0 +1,313 @@
+  }
+
+  TORCH_CHECK(
+      *target.min().data_ptr<int64_t>() >= -1, target_arg, " is out of range");
```
`target.min().item<int64_t>()` is more idiomatic

andreaskoepf replied:
ok!

andreaskoepf replied:
fixed in https://github.com/pytorch/pytorch/pull/28205/commits/833aca2ac06516b3541ab41acbb5f288dbf496c7

ezyang commented on aten/src/ATen/native/LossMultiLabelMargin.cpp:
```diff
@@ -0,0 +1,313 @@
+  auto target_contiguous = target.contiguous();
+
+  is_target.resize_as_(target);
+  TORCH_CHECK(is_target.is_contiguous(), "is_target must be contiguous");
```
Perhaps, test this before resizing?

andreaskoepf replied:
I though there might be cases in which resize_as_() makes the Tensor contiguous? If that happens it would be desired.

andreaskoepf committed:
Use item<T>() instead of *data_ptr<T>()

andreaskoepf commented:
References #24736, #24737

yf225 commented:
`caffe2-py2-cuda9.0-cudnn7-windows-build` on latest master seems to suggest that this PR breaks master:

```
18:22:45 ..\aten\src\ATen\native\LossMultiLabelMargin.cpp(201): error C2065: 'input_contiguous': undeclared identifier
18:22:45 ..\aten\src\ATen\native\LossMultiLabelMargin.cpp(261): error C2059: syntax error: ')'
18:22:45 ..\aten\src\ATen\native\LossMultiLabelMargin.cpp(201): error C2065: 'target_contiguous': undeclared identifier
18:22:45 ..\aten\src\ATen\native\LossMultiLabelMargin.cpp(201): error C2065: 'is_target_contiguous': undeclared identifier
18:22:45 ..\aten\src\ATen\native\LossMultiLabelMargin.cpp(201): error C2065: 'reduction': undeclared identifier
18:22:45 ..\aten\src\ATen\native\LossMultiLabelMargin.cpp(201): error C2065: 'nframe': undeclared identifier
18:22:45 ..\aten\src\ATen\native\LossMultiLabelMargin.cpp(201): error C2065: 'dim': undeclared identifier
18:22:45 ..\aten\src\ATen\native\LossMultiLabelMargin.cpp(201): error C2065: 'grad_input': undeclared identifier
18:22:45 ..\aten\src\ATen\native\LossMultiLabelMargin.cpp(201): error C2065: 'grad_output': undeclared identifier
18:22:45 ..\aten\src\ATen\native\LossMultiLabelMargin.cpp(261): error C3536: 'd': cannot be used before it is initialized
18:22:45 ..\aten\src\ATen\native\LossMultiLabelMargin.cpp(261): error C2059: syntax error: 'constant'
18:22:45 ..\aten\src\ATen\native\LossMultiLabelMargin.cpp(261): error C3536: 'grad_output_acc': cannot be used before it is initialized
18:22:45 ..\aten\src\ATen\native\LossMultiLabelMargin.cpp(261): error C2109: subscript requires array or pointer type
18:22:45 ..\aten\src\ATen\native\LossMultiLabelMargin.cpp(201): error C2065: 'is_target_arg': undeclared identifier
18:22:45 ..\aten\src\ATen\native\LossMultiLabelMargin.cpp(201): error C2440: '<function-style-cast>': cannot convert from 'initializer list' to 'c10::Error'
```
I am reverting it in the hope that it will fix master.

andreaskoepf commented:
@yf225 @ezyang I wonder why did 833aca2 then pass all 72 checks?

ezyang commented:
Caffe2 Windows isn't run on PRs because it is super flaky; we've been hitting Windows binary size limits for a while

ezyang commented:
FWIW, the error message here makes ZERO sense

andreaskoepf commented:
@yf225 Is it possible to see the first errors that were generated or the full build-log? Some of the errors look as if there was a general problem referencing variables outside the lambda expression. Others really look strange.

yf225 commented:
@andreaskoepf Yes this is the full build log: https://ci.pytorch.org/jenkins/job/caffe2-builds/job/py2-cuda9.0-cudnn7-windows-build/71651//consoleText

andreaskoepf commented:
@yf225 thanks Will! I can only speculate that VC does not like the long macro __VA_ARGS__ or has strange problems with TORCH_CHECK inside the macro. I think I have to setup a windows build environment to track this down.

andreaskoepf commented:
@yf225 I tried to compile PyTorch on Windows with VS2017 (cpu only) but both my branch and the current master version fail after successfully compiling several hundred files with other errors, e.g.:

```
C:\code\pytorch\aten\src\ATen/core/ATenDispatch.h(233): error C3520: 'value': parameter pack must be expanded in this context
C:\code\pytorch\aten\src\ATen/core/TensorMethods.h(67): note: see reference to function template instantiation 'Result at::ATenOpTable::callUnboxed<void,const at::Tensor&,const at::Tensor&,bool,bool>(const at::Tensor &,const at::Tensor &,bool,bool) const' being compiled
        with
        [
            Result=void
        ]
C:\code\pytorch\aten\src\ATen/core/ATenDispatch.h(234): error C2440: 'default argument': cannot convert from 'nullptr' to 'enable_if<_Test,_Ty>::type'
        with
        [
            _Ty=std::nullptr_t
        ]
C:\code\pytorch\aten\src\ATen/core/ATenDispatch.h(234): note: 'enable_if<_Test,_Ty>::type' may be a value type at runtime: consider using 'enable_if<_Test,_Ty>::type()' instead
        with
        [
            _Ty=std::nullptr_t
        ]
[2/1180] Building CXX object caffe2\CMakeFiles\torch.dir\__\aten\src\ATen\core\Tensor.cpp.obj
FAILED: caffe2/CMakeFiles/torch.dir/__/aten/src/ATen/core/Tensor.cpp.obj
```

`cl.exe` says it is Version 19.16.27034 for x64 ...

ezyang commented:
@andreaskoepf I think that's the same problem @fmassa was seeing, where some versions of MSVC don't correctly handle nested parameter packs. I still need to add a workaround, but if you want to repro we're going to have to find another version of MSVC until I get to this.
